tests/integration/logic.rs

```diff
@@ -0,0 +1,355 @@
+                    .map_err(|err| typedb_client::Error::Other(format!("{err:?}")))?
+                    .into_conjunction(),
+                typeql_lang::parse_variable("$x has is-still-good true")
+                    .map_err(|err| typedb_client::Error::Other(format!("{err:?}")))?,
+            ).await?;
+            transaction.logic().put_rule(
+                "all-milk-is-good".to_string(),
+                typeql_lang::parse_pattern("{ $x isa milk; }")
+                    .map_err(|err| typedb_client::Error::Other(format!("{err:?}")))?
+                    .into_conjunction(),
+                typeql_lang::parse_variable("$x has is-still-good true")
+                    .map_err(|err| typedb_client::Error::Other(format!("{err:?}")))?,
+            ).await?;
+            transaction.commit().await?;
+        }
+
+        let session = Session::new(databases.get(common::TEST_DATABASE).await?, Data).await?;
+        let transaction = session.transaction(Write).await?;
+        let data = r#"insert $x isa milk, has age-in-days 5;"#;
+        let _ = transaction.query().insert(data)?;
+        let data = r#"insert $x isa milk, has age-in-days 10;"#;
+        let _ = transaction.query().insert(data)?;
+        let data = r#"insert $x isa milk, has age-in-days 15;"#;
+        let _ = transaction.query().insert(data)?;
+        transaction.commit().await?;
+
+        let with_inference_and_explanation = Options::new().infer(true).explain(true);
+        let transaction = session.transaction_with_options(Read, with_inference_and_explanation).await?;
+        let answer_stream = transaction.query().match_(
+            r#"match $x has is-still-good $a;"#,
+        )?;
+        let answers = answer_stream.try_collect::<Vec<ConceptMap>>().await?;
+
+        assert_eq!(3, answers.len());
+
+        assert!(!answers.get(0).unwrap().explainables.is_empty());
+        assert!(!answers.get(1).unwrap().explainables.is_empty());
+        assert!(!answers.get(2).unwrap().explainables.is_empty());
+
+        let age_in_days = transaction.concept().get_attribute_type(String::from("age-in-days")).await?.unwrap();
+        for ans in answers {
+            match ans.map.get("x").unwrap() {
+                Concept::Entity(entity) => {
+                    let attributes: Vec<Attribute> = entity.get_has(&transaction, vec![age_in_days.clone()], vec![])?.try_collect().await?;
+                    if attributes.first().unwrap().value == Value::Long(15) {
+                        assert_single_explainable_explanations(&ans, 1, &transaction).await;
+                    } else {
+                        assert_single_explainable_explanations(&ans, 2, &transaction).await;
+                    }
+                },
+                _ => panic!("Incorrect Concept type: {:?}", ans.map.get("x").unwrap()),
+            }
+        }
+
+        Ok(())
+    }
+}
+
+async fn assert_single_explainable_explanations(
```

Comment:
This name doesn't tell us what it is that we're asserting about the explanations. We're checking here, a) that the count is as we expect, and b) that the concepts we receive in the explanation match the concepts in the original answer up to the projection. Which implies to me that this should be split into two functions:

```rust
// before
assert_single_explainable_explanations(&ans, 1, &transaction).await;

// after
let explanables = all_explainables(&ans);
assert_eq!(explainables.len(), 1);
assert!(explanation_concepts_match_projection(&ans, explainables[0]).await);
```

Reply:
a) - we have two count checks here.
I've extracted helper functions, but prefer to keep an old function to avoid code duplication. Renamed it, but the name is still not perfect.

Reply:
Fair enough. The name looks good.